staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/unstructured_conversion_test.go

```diff
@@ -0,0 +1,465 @@
+			outUnstructured := &unstructured.Unstructured{}
+			err := scheme.Convert(testCase.objectToConvert, outUnstructured, nil)
+			assert.Equal(t, testCase.expectedErr, err)
+			assert.Equal(t, testCase.expectedConvertedUnstructured, outUnstructured)
+		})
+	}
+}
+
+func TestUnstructuredToObjectConversion(t *testing.T) {
+	scheme, _ := test.TestScheme()
+	testCases := []struct {
+		name                    string
+		unstructuredToConvert   *unstructured.Unstructured
+		convertingObject        runtime.Object
+		expectPanic             bool
+		expectedErrFunc         func(err error) bool
+		expectedConvertedObject runtime.Object
+	}{
+		{
+			name:                  "convert nil unstructured to versioned object will panic",
+			unstructuredToConvert: nil,
+			convertingObject:      &testapigroupv1.Carp{},
+
+			expectPanic:             true,
+			expectedConvertedObject: &testapigroupv1.Carp{},
+		},
+		{
+			name: "convert empty unstructured w/o gvk to versioned object should fail",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{},
+			},
+			convertingObject: &testapigroupv1.Carp{},
+			expectedErrFunc: func(err error) bool {
+				return reflect.DeepEqual(err, runtime.NewMissingKindErr("unstructured object has no kind"))
+			},
+			expectedConvertedObject: &testapigroupv1.Carp{},
+		},
+		{
+			name: "convert empty versioned unstructured to versioned object should succeed",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+				},
+			},
+			convertingObject:        &testapigroupv1.Carp{},
+			expectedConvertedObject: &testapigroupv1.Carp{},
+		},
+		{
+			name: "convert empty unstructured w/o gvk to versioned object should fail",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{},
+			},
+			convertingObject: &testapigroupv1.Carp{},
+			expectedErrFunc: func(err error) bool {
+				return reflect.DeepEqual(err, runtime.NewMissingKindErr("unstructured object has no kind"))
+			},
+			expectedConvertedObject: &testapigroupv1.Carp{},
+		},
+		{
+			name: "convert valid versioned unstructured to versioned object should succeed",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+					"metadata": map[string]interface{}{
+						"creationTimestamp": nil,
+						"name":              "noxu",
+					},
+					"spec": map[string]interface{}{
+						"hostname": "example.com",
+					},
+					"status": map[string]interface{}{},
+				},
+			},
+			convertingObject: &testapigroupv1.Carp{},
+			expectedConvertedObject: &testapigroupv1.Carp{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "noxu",
+				},
+				Spec: testapigroupv1.CarpSpec{
+					Hostname: "example.com",
+				},
+			},
+		},
+		{
+			name: "convert valid versioned unstructured to internal object should succeed",
```

Comment:
this is very surprising to me... this means the conversion is not case-sensitive with respect to json tags (since the internal object is missing them entirely). I would have expected failure here